test/cpp/end2end/test_service_impl.cc

```diff
@@ -579,35 +591,43 @@ experimental::ServerUnaryReactor* CallbackTestServiceImpl::Echo(
       Finish(Status::OK);
     }
     void LoopUntilCancelled(int loop_delay_us) {
-      if (!ctx_->IsCancelled()) {
-        alarm_.experimental().Set(
-            gpr_time_add(gpr_now(GPR_CLOCK_REALTIME),
-                         gpr_time_from_micros(loop_delay_us, GPR_TIMESPAN)),
-            [this, loop_delay_us](bool ok) {
-              if (!ok) {
-                EXPECT_TRUE(ctx_->IsCancelled());
-              }
-              LoopUntilCancelled(loop_delay_us);
-            });
-      } else {
-        Finish(Status::CANCELLED);
+      {
+        std::lock_guard<std::mutex> l(alarm_mu_);
+        if (!ctx_->IsCancelled()) {
```

Comment:
Remind me why we have to set alarms and loop instead of using a condition variable

Reply:
We can't do a CV wait inside a reaction because that's a blocking operation. The other approach instead of looping would be to store a continuation function in the class object and then invoke that continuation function from OnCancel, or something like that.

Reply:
Recap from our offline discussion:
The test code would be a lot simpler if we
1) Use a condition variable that's set to true in OnCancel and delete the alarm use in this function
2) Rename LoopUntilCancelled to FinishWhenCancelled
3) Have FinishWhenCancelled run from another thread, and join that thread in OnDone

We could probably get rid of alarms throughout this file, but that's out of scope for this PR.